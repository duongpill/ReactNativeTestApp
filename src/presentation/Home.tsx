/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, { FC, useCallback, useEffect, useMemo, useState } from 'react';
import {
  ActivityIndicator,
  FlatList,
  Text,
  View,
} from 'react-native';
import styles from '../../styles/HomeStyles';
import PostCard from './components/PostCard';
import { DependencyInjections } from '../di/DependencyInjections';
import { SafeAreaView } from 'react-native-safe-area-context';
import HomeHeader from './components/HomeHeader';
import { Post } from '../domain/entity/Post';
import Loading from './components/Loading';

const Home: FC = () => {

  const [loading, setLoading] = useState(false);
  const [items, setItems] = useState<Post[]>([]);
  const [paginationToken, setPaginationToken] = useState<string>()

  useEffect(() => {
    fetchData()
  }, []);

  const fetchData = useCallback(async () => {
    if(loading) {
      return;
    }

    setLoading(true)
    var postResponse = await DependencyInjections.instance().getPostsUseCase(paginationToken)
    if(postResponse){
      setItems((currentItems) => [...currentItems, ...postResponse?.posts!])
      setPaginationToken(postResponse.paginationToken)
    }
    setLoading(false)
  }, [paginationToken]);

  const loadMore = () => {
    if(!loading){
      fetchData()
    }
  }

  const renderFooter = useMemo(() => {
    console.log('renderFooter')
    return (
      <View>
        <Text></Text>
        {(loading && items.length > 0) && <ActivityIndicator />}
      </View>
    )
  }, [items]);

  const handleEmpty = () => {
    console.log('handleEmpty')
    return !loading && <Text style={styles.textEmpty}>Data not found!</Text>;
  };

  return (
    <SafeAreaView style={styles.container}>
      <HomeHeader title="Home" />
      {(items.length == 0) && <Loading isLoading={loading} />}
      <View style={styles.flatListContainer}>
        <FlatList
          data={items}
          contentContainerStyle={{ gap: 10, flexGrow: 1 }}
          showsVerticalScrollIndicator={false}
          renderItem={({item}) => (
            <PostCard {...item} />
          )}
          keyExtractor={(index) => index.toString()}
          ListEmptyComponent={handleEmpty}
          ListFooterComponent={renderFooter}
          onEndReached={loadMore}
          onEndReachedThreshold={0.1}
        />
      </View>
    </SafeAreaView>
  );
}

export default Home;