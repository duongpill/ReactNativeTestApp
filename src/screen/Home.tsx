/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, { FC, useEffect, useState } from 'react';
import {
  ActivityIndicator,
  FlatList,
  Text,
  View,
} from 'react-native';
import styles from '../../styles/HomeStyles';
import { Post, PostResponse } from '../models/Post';
import PostCard from './components/PostCard';
import { Dependencies } from '../di/Dependencies';
import { SafeAreaView } from 'react-native-safe-area-context';
import HomeHeader from './components/HomeHeader';

const Home: FC = () => {

  const [loading, setLoading] = useState(false);
  const [items, setItems] = useState<Post[]>([]);
  const [paginationToken, setPaginationToken] = useState('');

  useEffect(() => {
    fetchData('')
  }, []);

  const fetchData = async(pagination_token: string) => {
    if(loading) {
      return;
    }

    var postResponse: PostResponse | null
    setLoading(true)
    if(pagination_token){
      postResponse = await Dependencies.instance.getPostRepository().loadMorePost(pagination_token)
    } else {
      postResponse = await Dependencies.instance.getPostRepository().loadPost()
    }
    if(postResponse){
      // console.log("Post: " + postResponse.data.items.length)
      setItems([...items, ...postResponse.data.items])
      setPaginationToken(postResponse.pagination_token)
    }
    setLoading(false)
  }

  // Function to handle load more (pagination)
  const loadMore = () => {
    // console.log('loadmore: ' + paginationToken + ' -- ' + loading)
    if (!loading) {
      fetchData(paginationToken);
    }
  };

  const renderFooter = () => {
    return (
      <View>
        <Text></Text>
        {(loading && items.length > 0) && <ActivityIndicator />}
      </View>
    )
  };

  const handleEmpty = () => {
    return !loading && <Text style={styles.textEmpty}>Data not found!</Text>;
  };

  return (
    <SafeAreaView style={styles.container}>
      <View>
        <HomeHeader title="Home"/>
        <View style={styles.flatListContainer}>
          <FlatList
            data={items}
            contentContainerStyle={{ gap: 10, flexGrow: 1 }}
            showsVerticalScrollIndicator={false}
            renderItem={({item, index}) => (
              <PostCard 
                {...item}
              />
            )}
            keyExtractor={(item, index) => index.toString()}
            ListEmptyComponent={handleEmpty}
            ListFooterComponent={renderFooter}
            onEndReached={loadMore}
            onEndReachedThreshold={0.1}
          />
        </View>
      </View>
    </SafeAreaView>
  );
}

export default Home;