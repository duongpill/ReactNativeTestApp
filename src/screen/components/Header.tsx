import React, { FC } from "react";
import { Image, Text, TouchableOpacity, View } from "react-native";
import styles from "../../../styles/HeaderStyles";
import { useNavigation } from "@react-navigation/native";

interface HeaderProps {
    title?: string;
}

const Header: FC<HeaderProps> = ({title}) => {
    const navigation = useNavigation();
    return (
        <TouchableOpacity testID="headerBackButton" style={styles.navContainer} onPress={() => navigation.goBack()}>
            <Image
                source={require('../../assets/ic_back.png')}
                style={styles.navBack}
            />
            <Text style={styles.navText}>{title}</Text>
        </TouchableOpacity>
    )
}

export default Header;