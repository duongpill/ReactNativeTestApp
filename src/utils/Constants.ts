import { BufferConfig, ResizeMode, SelectedVideoTrackType } from "react-native-video";

export const API = {
    url: 'https://instagram-scraper-api2.p.rapidapi.com',
    headers: {
        'x-rapidapi-key': 'f7a3ef1d49mshc8f5be43cb0ba30p156b3ejsne7083b61cf72',
        'x-rapidapi-host': 'instagram-scraper-api2.p.rapidapi.com'
    }
}

export const defaultValue = {
    rate: 1,
    volume: 1,
    muted: false,
    resizeMode: ResizeMode.CONTAIN,
    duration: 0.0,
    currentTime: 0.0,
    videoWidth: 0,
    videoHeight: 0,
    paused: false,
    fullscreen: true,
    decoration: true,
    isLoading: false,
    audioTracks: [],
    textTracks: [],
    videoTracks: [],
    selectedAudioTrack: undefined,
    selectedTextTrack: undefined,
    selectedVideoTrack: {
      type: SelectedVideoTrackType.AUTO,
    },
    srcListId: 0,
    loop: false,
    showRNVControls: false,
    useCache: false,
    poster: undefined,
    showNotificationControls: false,
    isSeeking: false,
  };

export const bufferConfig: BufferConfig = {
    minBufferMs: 15000,
    maxBufferMs: 50000,
    bufferForPlaybackMs: 2500,
    bufferForPlaybackAfterRebufferMs: 5000,
    live: {
      targetOffsetMs: 500,
    },
}