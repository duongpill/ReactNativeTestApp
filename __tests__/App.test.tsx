/**
 * @format
 */

import { BackHandler, Dimensions, Text, TouchableOpacity, View } from 'react-native';
import React from 'react';

// Note: import explicitly to use the types shipped with jest.
import {describe, expect, it, jest, test} from '@jest/globals';

import Header from '../src/presentation/components/Header';
import { act, fireEvent, render, waitFor, waitForElementToBeRemoved } from '@testing-library/react-native';

import { API } from '../src/utils/Constants';
import Home from '../src/presentation/Home';

const mockedDispatch = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
      ...actualNav,
      useNavigation: () => ({ 
          navigate: jest.fn(),
          dispatch: jest.fn(),
          goBack: mockedDispatch,
      }),
  };
});

test('Header renders correctly', async () => {
  const header = render(<Header/>);
  await waitFor(() => { 
    expect(header).toMatchSnapshot();
  });
});

test('Home renders correctly', async () => {
  const home = render(<Home />);
  await waitFor(() => { 
    expect(home).toMatchSnapshot();
  });
});

describe('Header back button', () => {
  beforeEach(() => {
    mockedDispatch.mockClear();
  });

  it("Header back button press", () => {
    const { getByTestId } = render(<Header />);  
  
    const button = getByTestId('headerBackButton')
    
    fireEvent.press(button);

    expect(button).toBeOnTheScreen();
    expect(mockedDispatch).toHaveBeenCalledTimes(1);
  });
});

describe('Try to fetch posts', () => {

  const getProducts = async () => {
    var url = `${API.url}/v1.2/posts?username_or_id_or_url=mrbeast`;
    const options = {
      method: 'GET',
      headers: API.headers
    };
    const res = await fetch(url, options);
    const text = await res.text();
    return text ? JSON.stringify(text) : [];
  }
  
  const unmockedFetch = global.fetch;

  beforeAll(() => {
    global.fetch = jest.fn(() => Promise.resolve(new Response()));
  })

  afterAll(() => {
    global.fetch = unmockedFetch;
  })

  test('getPost success', async () => {
    const result = await getProducts();
    expect(Array.isArray(result)).toEqual(true);
  })

  test('getPost fails', async () => {
    const result = await getProducts();
    expect(result.length).toEqual(0);
  })
});