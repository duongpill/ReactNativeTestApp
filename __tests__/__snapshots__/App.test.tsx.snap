// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "height": 60,
      "opacity": 1,
      "width": "100%",
    }
  }
  testID="headerBackButton"
>
  <Image
    source={
      {
        "testUri": "../../../src/assets/ic_back.png",
      }
    }
    style={
      {
        "height": 22,
        "padding": 10,
        "tintColor": "#000",
        "width": 22,
      }
    }
  />
  <Text
    style={
      {
        "color": "#444",
        "fontSize": 26,
        "fontWeight": "700",
      }
    }
  />
</View>
`;

exports[`Home renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f0f0f0",
      "flex": 1,
      "minHeight": 1334,
      "paddingHorizontal": 15,
    }
  }
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "space-between",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#444",
            "fontSize": 26,
            "fontWeight": "700",
          }
        }
      >
        Home
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            {
              "testUri": "../../../src/assets/search.png",
            }
          }
          style={
            {
              "alignItems": "flex-end",
              "height": 22,
              "padding": 10,
              "tintColor": "#000",
              "width": 22,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <RCTScrollView
        ListEmptyComponent={[Function]}
        ListFooterComponent={[Function]}
        contentContainerStyle={
          {
            "flexGrow": 1,
            "gap": 10,
          }
        }
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onEndReached={[Function]}
        onEndReachedThreshold={0.1}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onLayout={[Function]}
          >
            <View>
              <Text />
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaView>
`;
